segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,450 @@
+                        }
+                        BufferedReader reader = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(contents)));
+                        String line;
+
+                        while ((line = reader.readLine()) != null) {
+                            // Handle only whole records.
+                            if (line.startsWith(START_TOKEN) && line.endsWith(END_TOKEN)) {
+                                String[] records = line.split(RECORD_SEPARATOR);
+                                for (String record : records) {
+                                    String[] parts = record.split(":");
+                                    if (ADD_RECORD.equals(parts[0]) && parts.length == 5) {
+                                        String segmentName = parts[1];
+                                        String oldChunkName = parts[2];
+                                        String newChunkName = parts[3];
+                                        long offset = Long.parseLong(parts[4]);
+
+                                        applyChunkAddition(txn, chunkStartOffsets, segmentName, oldChunkName, newChunkName, offset);
+                                    }
+                                    if (TRUNCATE_RECORD.equals(parts[0]) && parts.length == 5) {
+                                        String segmentName = parts[1];
+                                        long truncateAt = Long.parseLong(parts[2]);
+                                        String firstChunkName = parts[3];
+                                        long truncateStartAt = Long.parseLong(parts[4]);
+                                        finalTruncateOffsets.put(segmentName, truncateAt);
+                                        finalFirstChunkStartsAtOffsets.put(segmentName, truncateStartAt);
+                                    }
+                                }
+                            }
+                        }
+                        // Move to next file.
+                        fileIndexToRecover++;
+                    }
+                }
+
+                // Step 3: Adjust the length of the last chunk.
+                for (String systemSegment : systemSegments) {
+                    SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(systemSegment);
+                    segmentMetadata.checkInvariants();
+                    if (null != segmentMetadata.getLastChunk()) {
+                        val chunkInfo = chunkStorage.getInfo(segmentMetadata.getLastChunk());
+                        long length = chunkInfo.getLength();
+
+                        ChunkMetadata lastChunk = (ChunkMetadata) txn.get(segmentMetadata.getLastChunk());
+                        Preconditions.checkState(null != lastChunk);
+                        lastChunk.setLength(Math.toIntExact(length));
+                        txn.update(lastChunk);
+                        segmentMetadata.setLength(segmentMetadata.getLength() + length);
```

Comment:
In the presence of zombie segment store, the new and old instances are running  concurrently. The length of last chunk is held in memory of zombie process and not written out to file. The new instance knows what is the last chunk but not its current length. This is why we check the length directly and update. Now the zombie instance might continue to write even after this but that does not matter because in the view of new instance that chunk ended at offset new instance decides to save. The rule here is that the new instance always writes to a new chunk.